src/fuelTanks/CCPACSVessel.cpp

```diff
@@ -0,0 +1,725 @@
+    double area              = 0.;
+
+    TopTools_IndexedMapOfShape shapeMap;
+    TopExp::MapShapes(shape->Shape(), TopAbs_FACE, shapeMap);
+    for (int i = 1; i <= shapeMap.Extent(); ++i) {
+        const std::string& faceName = shape->GetFaceTraits(i - 1).Name();
+        if (faceName != "symmetry") {
+            const TopoDS_Face& face = TopoDS::Face(shapeMap(i));
+            GProp_GProps shapeProps;
+            BRepGProp::SurfaceProperties(face, shapeProps);
+            area += shapeProps.Mass();
+        }
+    }
+
+    return area;
+}
+
+double CCPACSVessel::GetCircumference(const int segmentIndex, const double eta)
+{
+    if (m_segments_choice1) {
+        return static_cast<CCPACSFuselageSegment&>(GetSegment(segmentIndex)).GetCircumference(eta);
+    }
+    else {
+        throw CTiglError(_vesselTypeException);
+    }
+}
+
+bool CCPACSVessel::IsVesselViaSegments() const
+{
+    if (m_segments_choice1) {
+        return true;
+    }
+    else if (m_domeType_choice2) {
+        return false;
+    }
+    else {
+        throw CTiglError("Vessel type could not be determined.");
+    }
+}
+
+bool CCPACSVessel::IsVesselViaDesignParameters() const
+{
+    if (m_segments_choice1) {
+        return false;
+    }
+    else if (m_domeType_choice2) {
+        return true;
+    }
+    else {
+        throw CTiglError("Vessel type could not be determined.");
+    }
+}
+
+bool CCPACSVessel::HasSphericalDome() const
+{
+    return GetEllipsoidDome() && GetEllipsoidDome()->GetHalfAxisFraction() == 1.0;
+}
+
+bool CCPACSVessel::HasEllipsoidDome() const
+{
+    return GetEllipsoidDome() != nullptr;
+}
+
+bool CCPACSVessel::HasIsotensoidDome() const
+{
+    return GetIsotensoidDome() != nullptr;
+}
+
+bool CCPACSVessel::HasTorisphericalDome() const
+{
+    return GetTorisphericalDome() != nullptr;
+}
+
+CCPACSEllipsoidDome const* CCPACSVessel::GetEllipsoidDome() const
+{
+    if (m_domeType_choice2) {
+        return m_domeType_choice2->GetEllipsoid_choice1().get_ptr();
+    }
+    return nullptr;
+}
+
+CCPACSTorisphericalDome const* CCPACSVessel::GetTorisphericalDome() const
+{
+    if (m_domeType_choice2) {
+        return m_domeType_choice2->GetTorispherical_choice2().get_ptr();
+    }
+    return nullptr;
+}
+
+CCPACSIsotensoidDome const* CCPACSVessel::GetIsotensoidDome() const
+{
+    if (m_domeType_choice2) {
+        return m_domeType_choice2->GetIsotensoid_choice3().get_ptr();
+    }
+    return nullptr;
+}
+
+CCPACSVessel::IsotensoidContour::IsotensoidContour(double rCyl, double rPolarOpening, int nodeNumber)
+{
+    double dPhi = 1.0 / nodeNumber;
+
+    radii.push_back(rCyl);
+    axialPositions.push_back(0.0);
+
+    double phi   = 0.0;
+    double alpha = std::asin(rPolarOpening / rCyl);
+
+    const double eps = std::numeric_limits<double>::epsilon();
+
+    while (std::tan(alpha) * std::tan(alpha) < 2 && radii.back() > 1.22 * rPolarOpening &&
+           (radii.size() < 2 || radii.back() < radii[radii.size() - 2] + eps)) {
```

Comment:
This condition in the second line translates to

```suggestion
           dr < -eps) {
```
I believe, which I find much easier to read. It just means terminate, once the radii stop decreasing. Numerically, eps is the *relative* machine precision, so conceptually it should be scaled by `radi.back()` in the tolerance. But even then, this is an extremely small tolerance.

We need to predefine `dr` outside the `while`-loop to something that satisfies this, e.g. `double dr = -1.`.

Reply:
To my understanding, this condition was only added to make sure that the while-loop terminates. Now mathematically, I cannot say for sure that this really is the case. In addition, it might take A LOT of iterations, before this second condition is met and the while-loop terminates and if it does, the first conditions aren't met. Doesn't this mean, that there was some error in the sequence and we should raise an exception?


```cpp
double dr = -1.0;
int iteration = 0;
int max_iter = 1e3;

while (std::tan(alpha) * std::tan(alpha) < 2 && radii.back() > 1.22 * rPolarOpening) {

   /* ... */

    iteration = iteration + 1;
    if ( dr > - eps*radii.back() || iteration > max_iter )  {
        // radius is not decreasing anymore or maximum number of iterations reached
        throw CTiglError("Couldn't construct isotensoid contour of vessel for the input parameters....");
    }
}
```

Reply:
```cpp
CCPACSVessel::IsotensoidContour::IsotensoidContour(double rCyl, double rPolarOpening, int nodeNumber)
{
    double dPhi = 1.0 / nodeNumber;

    radii.push_back(rCyl);
    axialPositions.push_back(0.0);

    // initial values
    double phi           = 0.0;
    double dr            = 0.0;
    double dx            = 1.0;
    double radius        = rCyl;
    double axialPosition = 0.0;

    double alpha = std::asin(rPolarOpening / rCyl);

    const double eps = 1e-8 * std::max(rCyl, rPolarOpening);
    int i            = 0;

    while (std::tan(alpha) * std::tan(alpha) < 2 && radius > 1.22 * rPolarOpening && dr >= 0.0) {

        radii.push_back(radius);
        axialPositions.push_back(axialPosition);

        phi   = phi + dPhi;
        alpha = std::asin(rPolarOpening / radii.back());

        double meridianRadius = radius / (std::cos(phi) * (2 - std::tan(alpha) * std::tan(alpha)));
        dr                    = meridianRadius * dPhi * std::sin(phi);
        radius                = radius - dr;
        dx                    = meridianRadius * dPhi * std::cos(phi);
        axialPosition         = axialPosition + dx;
    }

    // Add 5 points linear until polar opening
    double lastX = axialPositions.back();
    double lastR = radii.back();
    double prevX = axialPositions[axialPositions.size() - 2];
    double prevR = radii[radii.size() - 2];

    double slope = (lastR - prevR) / (lastX - prevX);

    double dxTotal = (rPolarOpening - lastR) / slope;
    double deltaX  = dxTotal / 4.0;
    double deltaR  = (rPolarOpening - lastR) / 4.0;

    axialPositions.reserve(axialPositions.size() + 4);
    radii.reserve(radii.size() + 4);

    for (int i = 1; i <= 4; ++i) {
        axialPositions.push_back(lastX + i * deltaX);
        radii.push_back(lastR + i * deltaR);
    }

    // Shifting to x=0
    double totalDx = axialPositions.back() - axialPositions.front();
    for (double& v : axialPositions) {
        v -= totalDx;
    }
}
```

Not sure if it is the most elegant approach, but it seems to reproduce the Python code well for the values I tested. Note that I chose `const double eps = 1e-8 * std::max(rCyl, rPolarOpening);`.

![isotensoids](https://github.com/user-attachments/assets/8af8d410-3faa-4953-bf27-538b54af6b07)